crates/history/src/query.rs

```diff
@@ -0,0 +1,120 @@
+//! # Encoding and decoding strategies for query strings.
+//!
+//! There are various strategies to map Rust types into HTTP query strings. The [`FromQuery`] and
+//! [`IntoQuery`] encode the logic for how this encoding and decoding is performed. These traits
+//! are public as a form of dependency inversion, so that you can override the decoding and
+//! encoding strategy being used.
+//!
+//! These traits are used by the [`History`](crate::History) trait, which allows for modifying the
+//! history state, and the [`Location`](crate::Location) struct, which allows for extracting the
+//! current location (and this query).
+//!
+//! ## Default Strategy
+//!
+//! By default, any Rust type that implements [`Serialize`] or [`Deserialize`](serde::Deserialize)
+//! has an implementation of [`IntoQuery`] or [`FromQuery`], respectively. This implementation uses
+//! the `serde_urlencoded` crate, which implements a standards-compliant `x-www-form-urlencoded`
+//! encoder and decoder. Some patterns are not supported by this crate, for example it is not
+//! possible to serialize arrays at the moment. If this is an issue for you, consider using the
+//! `serde_qs` crate.
+//!
+//! Example:
+//!
+//! ```rust,no_run
+//! use serde::{Serialize, Deserialize};
+//! use gloo_history::{MemoryHistory, History};
+//!
+//! #[derive(Serialize)]
+//! struct Query {
+//!     name: String,
+//! }
+//!
+//! let query = Query {
+//!     name: "user".into(),
+//! };
+//!
+//! let history = MemoryHistory::new();
+//! history.push_with_query("index.html", &query).unwrap();
+//! ```
+//!
+//! ## Other Strategies
+//!
+//! Depending on the features enabled, this module defines alternative strategies using wrapper
+//! types. These wrapper types can be used to wrap your struct to modify how it is encoded and
+//! decode. See the documentation for the structs in this module for further information.
+//!
+//! ## Custom Strategy
+//!
+//! If desired, the [`FromQuery`] and [`IntoQuery`] traits can also be manually implemented on
+//! types to customize the encoding and decoding strategies. See the documentation for these traits
+//! for more detail on how this can be done.
+use crate::error::HistoryError;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    borrow::{Borrow, Cow},
+    convert::Infallible,
+};
+
+/// Type that can be encoded into a query string.
+pub trait IntoQuery {
+    /// Error that can be returned from the conversion.
+    type Error;
+
+    /// Method to encode the query into a string.
+    fn encode(&self) -> Result<Cow<'_, str>, Self::Error>;
+}
+
+/// Type that can be decoded from a query string.
+pub trait FromQuery {
+    /// Target type after parsing.
+    type Target;
+    /// Error that can occur while parsing.
+    type Error;
+
+    /// Decode this query string into the target type.
+    fn decode(query: &str) -> Result<Self::Target, Self::Error>;
```

Comment:
I think the issue with `into_query` is that as you mentioned earlier, `Into` usually suggests something that cannot fail. Therefore it would need to be named `try_into_query()` and `try_from_query()`. But that point, I would rather suggest to rename the trait into `DecodeQuery` and `EncodeQuery`, because semantially the idea of decoding and encoding already suggests that it is an operation which can fail.

What do you think about that idea?

Reply:
I am fine with both solutions as `from` / `to` is not strictly interpreted as infallible in std. E.g.: `FromStr::from_str` is defined as `fn from_str(s: &str) -> Result<Self, Self::Err>;` and `OsStr::to_str` is defined as `pub fn to_str(&self) -> Option<&str>`.

> But that point, I would rather suggest to rename the trait into DecodeQuery and EncodeQuery, because semantially the idea of decoding and encoding already suggests that it is an operation which can fail.

It might make sense to combine decode and encode side of queries into `Query` trait as a query type in this case usually is used as a type to decode / encode the actual underlying query so it should implement both operations.

Reply:
I have taken both of your suggestions and renamed the traits to `FromQuery`/`ToQuery` and the methods to `from_query` and `to_query`.

> It might make sense to combine decode and encode side of queries into Query trait as a query type in this case usually is used as a type to decode / encode the actual underlying query so it should implement both operations.

Well, the two operations are different (encoding vs decoding). We do not want to force users of the API to implement both just because they want to use the `push_with_query()` API. That would not be a good design.

We can always have a supertrait:

```rust
pub trait Query {
    type Decoded;
    type DecodeError;
    type EncodeError;
    fn from_query()
    fn to_query()
}

impl<T: FromQuery + ToQuery> Query for T {
    type Decoded = <T as FromQuery>::Target;
    // ...
}
```

But we do not have a use for this type, as we have no method that requires **both** the ability to encode and to decode. All the APIs require either one but not both. That is why I think it is good that these are separate traits.